aiida/tools/graph/age_rules.py

```diff
@@ -0,0 +1,406 @@
+from collections import defaultdict
+import numpy as np
+
+from aiida.common.exceptions import InputValidationError
+from aiida.orm.querybuilder import QueryBuilder
+from aiida.tools.graph.age_entities import Basket
+
+
+def check_if_basket(entity_set):
+    if not isinstance(entity_set, Basket):
+        raise TypeError('You need to set the walkers with an AiidaEntitySet')
+
+
+class Operation(metaclass=ABCMeta):
+    """Operation metaclass - Base class for all AGE explorer classes"""
+
+    def __init__(self, max_iterations, track_edges):
+        self.set_max_iterations(max_iterations)
+        self._track_edges = track_edges  # Logical
+        self._iterations_done = None
+
+    def set_max_iterations(self, max_iterations):
+        """Set max iterations"""
+        if max_iterations is np.inf:
+            pass
+        elif not isinstance(max_iterations, int):
+            raise TypeError('max iterations has to be an integer')
+        self._maxiter = max_iterations
+
+    def get_iterations_done(self):
+        return self._iterations_done
+
+    @abstractmethod
+    def run(self, operational_set):
+        pass
+
+
+class QueryRule(Operation, metaclass=ABCMeta):
+    """
+    The abstract class QueryRule is the parent class for every rule that implements
+    a query (currently UpdateRule and ReplaceRule).
+    """
+
+    def __init__(self, querybuilder, max_iterations=1, track_edges=False):
+        """
+        :param querybuilder: An instance of the QueryBuilder class
+        :param int max_iterations: The number of iterations to run this query on
+        :param bool track_edges: Whether to track which edges are traversed and store them
+        """
+
+        def get_spec_from_path(queryhelp, idx):
+            if (
+                queryhelp['path'][idx]['entity_type'].startswith('node') or
+                queryhelp['path'][idx]['entity_type'].startswith('data') or
+                queryhelp['path'][idx]['entity_type'].startswith('process') or
+                queryhelp['path'][idx]['entity_type'] == ''
+            ):
+                # not really nice, this way, how about using
+                # from aiida.orm.utils.node import is_valid_node_type_string and checking entity_type?
+                result = 'nodes'
+            elif queryhelp['path'][idx]['entity_type'] == 'group':
+                result = 'groups'
+            else:
+                raise Exception('not understood entity from ( {} )'.format(queryhelp['path'][idx]['entity_type']))
+            return result
+
+        queryhelp = querybuilder.queryhelp
+
+        # Check if there is any projection:
+        query_projections = queryhelp['project']
+        for projection_key in query_projections:
+            if query_projections[projection_key] != []:
+                error_message = 'The input querybuilder must not have any projections in it.\n'
+                error_message = error_message + 'Instead, it has the following:\n'
```

Comment:
```
error_message = error_message + 'Instead, it has the following
```
can become 
```
error_message += 'Instead, it has the following
```
and thinking of it, is it not just better to use a single string interpolation?

Reply:
So, I've made an example for both versions: just using the "addinplace" (a) and using string interpolation (b). Personally I still like more the (a) version because at a glance I understand it much better than the other option, but this might just be me not being used to a more 'pythonic' style (or also I might not have chosen the best formating for this one?).

**Version a:**
```
if query_projections[projection_key] != []:
    error_message = 'The input querybuilder must not have any projections.\n'
    error_message += 'Instead, it has the following:\n'
    error_message += ' - Key: {}\n'.format(projection_key)
    error_message += ' - Val: {}\n'.format(query_projections[projection_key])
    raise ValueError(error_message)
```

**Version b:**
```
if query_projections[projection_key] != []:
    raise ValueError(
        'The input querybuilder must not have any projections in it.\n'
        'Instead, it has the following:\n - Key: {}\n - Val: {}\n'.format(
            projection_key, query_projections[projection_key]
        ))
```

If  you still have other reasons to prefer b (even if it is just to keep it consistent with the rest of the codebase) I will proceed with that, no problem.

Reply:
Let's keep b to stay consistent with the rest of the code